// ngx-push-notification/esm5/lib/ngx-push-notification.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import * as webNotification from 'simple-web-notification';
import * as i0 from "@angular/core";
var NgxPushNotificationService = /** @class */ (function () {
    function NgxPushNotificationService() {
    }
    /**
     * @param {?} options
     * @return {?}
     */
    NgxPushNotificationService.prototype.showNotification = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var observable = new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        function (observer) {
            webNotification.showNotification(options.title, {
                body: options.body,
                icon: options.icon,
                onClick: (/**
                 * @return {?}
                 */
                function onNotificationClicked() {
                    observer.next({
                        type: 'click'
                    });
                }),
                autoClose: 4000
            }, (/**
             * @param {?} error
             * @param {?} hide
             * @return {?}
             */
            function onShow(error, hide) {
                if (error) {
                    observer.error(error.message);
                }
                else {
                    observer.next({
                        type: 'show'
                    });
                    setTimeout((/**
                     * @return {?}
                     */
                    function hideNotification() {
                        observer.next({
                            type: 'close'
                        });
                        hide();
                    }), 4000);
                }
            }));
        }));
        return observable;
    };
    NgxPushNotificationService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ NgxPushNotificationService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function NgxPushNotificationService_Factory() { return new NgxPushNotificationService(); }, token: NgxPushNotificationService, providedIn: "root" });
    return NgxPushNotificationService;
}());
export { NgxPushNotificationService };
