// angular-archwizard/bundles/angular-archwizard.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('angular-archwizard', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = global || self, factory(global['angular-archwizard'] = {}, global.ng.core, global.ng.common));
}(this, function (exports, core, common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awWizardStepSymbol` directive can be used as an alternative to the `navigationSymbol` input of a [[WizardStep]]
     * to define the step symbol inside the navigation bar.  This way step symbol may contain arbitrary content.
     *
     * ### Syntax
     *
     * ```html
     * <ng-template awWizardStepSymbol>
     *     ...
     * </ng-template>
     * ```
     */
    var WizardStepSymbolDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param templateRef A reference to the content of the `ng-template` that contains this [[WizardStepSymbolDirective]]
         */
        function WizardStepSymbolDirective(templateRef) {
            this.templateRef = templateRef;
        }
        WizardStepSymbolDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'ng-template[awStepSymbol], ng-template[awWizardStepSymbol]'
                    },] }
        ];
        /** @nocollapse */
        WizardStepSymbolDirective.ctorParameters = function () { return [
            { type: core.TemplateRef }
        ]; };
        return WizardStepSymbolDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awWizardStepTitle` directive can be used as an alternative to the `stepTitle` input of a [[WizardStep]]
     * to define the content of a step title inside the navigation bar.
     * This step title can be freely created and can contain more than only plain text
     *
     * ### Syntax
     *
     * ```html
     * <ng-template awWizardStepTitle>
     *     ...
     * </ng-template>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardStepTitleDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param templateRef A reference to the content of the `ng-template` that contains this [[WizardStepTitleDirective]]
         */
        function WizardStepTitleDirective(templateRef) {
            this.templateRef = templateRef;
        }
        WizardStepTitleDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'ng-template[awStepTitle], ng-template[awWizardStepTitle]'
                    },] }
        ];
        /** @nocollapse */
        WizardStepTitleDirective.ctorParameters = function () { return [
            { type: core.TemplateRef }
        ]; };
        return WizardStepTitleDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Basic functionality every type of wizard step needs to provide
     *
     * @author Marc Arndt
     * @abstract
     */
    var WizardStep = /** @class */ (function () {
        function WizardStep() {
            /**
             * A symbol property, which contains an optional symbol for the step inside the navigation bar.
             * Takes effect when `stepSymbolTemplate` is not defined or null.
             */
            this.navigationSymbol = { symbol: '' };
            /**
             * A boolean describing if the wizard step is currently selected
             */
            this.selected = false;
            /**
             * A boolean describing if the wizard step has been completed
             */
            this.completed = false;
            /**
             * A boolean describing if the wizard step is shown as completed when the wizard is presented to the user
             *
             * Users will typically use `CompletedStepDirective` to set this flag
             */
            this.initiallyCompleted = false;
            /**
             * A boolean describing if the wizard step is being edited after being competed
             *
             * This flag can only be true when `selected` is true.
             */
            this.editing = false;
            /**
             * A boolean describing, if the wizard step should be selected by default, i.e. after the wizard has been initialized as the initial step
             */
            this.defaultSelected = false;
            /**
             * A boolean describing if the wizard step is an optional step
             */
            this.optional = false;
            /**
             * A function or boolean deciding, if this step can be entered
             */
            this.canEnter = true;
            /**
             * A function or boolean deciding, if this step can be exited
             */
            this.canExit = true;
            /**
             * This [[EventEmitter]] is called when the step is entered.
             * The bound method should be used to do initialization work.
             */
            this.stepEnter = new core.EventEmitter();
            /**
             * This [[EventEmitter]] is called when the step is exited.
             * The bound method can be used to do cleanup work.
             */
            this.stepExit = new core.EventEmitter();
        }
        Object.defineProperty(WizardStep.prototype, "hidden", {
            /**
             * Returns if this wizard step should be visible to the user.
             * If the step should be visible to the user false is returned, otherwise true
             */
            get: /**
             * Returns if this wizard step should be visible to the user.
             * If the step should be visible to the user false is returned, otherwise true
             * @return {?}
             */
            function () {
                return !this.selected;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * This method returns true, if this wizard step can be transitioned with a given direction.
         * Transitioned in this case means either entered or exited, depending on the given `condition` parameter.
         *
         * @param condition A condition variable, deciding if the step can be transitioned
         * @param direction The direction in which this step should be transitioned
         * @returns A [[Promise]] containing `true`, if this step can transitioned in the given direction
         * @throws An `Error` is thrown if `condition` is neither a function nor a boolean
         */
        /**
         * This method returns true, if this wizard step can be transitioned with a given direction.
         * Transitioned in this case means either entered or exited, depending on the given `condition` parameter.
         *
         * @throws An `Error` is thrown if `condition` is neither a function nor a boolean
         * @private
         * @param {?} condition A condition variable, deciding if the step can be transitioned
         * @param {?} direction The direction in which this step should be transitioned
         * @return {?} A [[Promise]] containing `true`, if this step can transitioned in the given direction
         */
        WizardStep.canTransitionStep = /**
         * This method returns true, if this wizard step can be transitioned with a given direction.
         * Transitioned in this case means either entered or exited, depending on the given `condition` parameter.
         *
         * @throws An `Error` is thrown if `condition` is neither a function nor a boolean
         * @private
         * @param {?} condition A condition variable, deciding if the step can be transitioned
         * @param {?} direction The direction in which this step should be transitioned
         * @return {?} A [[Promise]] containing `true`, if this step can transitioned in the given direction
         */
        function (condition, direction) {
            if (typeof (condition) === typeof (true)) {
                return Promise.resolve((/** @type {?} */ (condition)));
            }
            else if (condition instanceof Function) {
                return Promise.resolve(condition(direction));
            }
            else {
                return Promise.reject(new Error("Input value '" + condition + "' is neither a boolean nor a function"));
            }
        };
        /**
         * A function called when the step is entered
         *
         * @param direction The direction in which the step is entered
         */
        /**
         * A function called when the step is entered
         *
         * @param {?} direction The direction in which the step is entered
         * @return {?}
         */
        WizardStep.prototype.enter = /**
         * A function called when the step is entered
         *
         * @param {?} direction The direction in which the step is entered
         * @return {?}
         */
        function (direction) {
            this.stepEnter.emit(direction);
        };
        /**
         * A function called when the step is exited
         *
         * @param direction The direction in which the step is exited
         */
        /**
         * A function called when the step is exited
         *
         * @param {?} direction The direction in which the step is exited
         * @return {?}
         */
        WizardStep.prototype.exit = /**
         * A function called when the step is exited
         *
         * @param {?} direction The direction in which the step is exited
         * @return {?}
         */
        function (direction) {
            this.stepExit.emit(direction);
        };
        /**
         * This method returns true, if this wizard step can be entered from the given direction.
         * Because this method depends on the value `canEnter`, it will throw an error, if `canEnter` is neither a boolean
         * nor a function.
         *
         * @param direction The direction in which this step should be entered
         * @returns A [[Promise]] containing `true`, if the step can be entered in the given direction, false otherwise
         * @throws An `Error` is thrown if `anEnter` is neither a function nor a boolean
         */
        /**
         * This method returns true, if this wizard step can be entered from the given direction.
         * Because this method depends on the value `canEnter`, it will throw an error, if `canEnter` is neither a boolean
         * nor a function.
         *
         * @throws An `Error` is thrown if `anEnter` is neither a function nor a boolean
         * @param {?} direction The direction in which this step should be entered
         * @return {?} A [[Promise]] containing `true`, if the step can be entered in the given direction, false otherwise
         */
        WizardStep.prototype.canEnterStep = /**
         * This method returns true, if this wizard step can be entered from the given direction.
         * Because this method depends on the value `canEnter`, it will throw an error, if `canEnter` is neither a boolean
         * nor a function.
         *
         * @throws An `Error` is thrown if `anEnter` is neither a function nor a boolean
         * @param {?} direction The direction in which this step should be entered
         * @return {?} A [[Promise]] containing `true`, if the step can be entered in the given direction, false otherwise
         */
        function (direction) {
            return WizardStep.canTransitionStep(this.canEnter, direction);
        };
        /**
         * This method returns true, if this wizard step can be exited into given direction.
         * Because this method depends on the value `canExit`, it will throw an error, if `canExit` is neither a boolean
         * nor a function.
         *
         * @param direction The direction in which this step should be left
         * @returns A [[Promise]] containing `true`, if the step can be exited in the given direction, false otherwise
         * @throws An `Error` is thrown if `canExit` is neither a function nor a boolean
         */
        /**
         * This method returns true, if this wizard step can be exited into given direction.
         * Because this method depends on the value `canExit`, it will throw an error, if `canExit` is neither a boolean
         * nor a function.
         *
         * @throws An `Error` is thrown if `canExit` is neither a function nor a boolean
         * @param {?} direction The direction in which this step should be left
         * @return {?} A [[Promise]] containing `true`, if the step can be exited in the given direction, false otherwise
         */
        WizardStep.prototype.canExitStep = /**
         * This method returns true, if this wizard step can be exited into given direction.
         * Because this method depends on the value `canExit`, it will throw an error, if `canExit` is neither a boolean
         * nor a function.
         *
         * @throws An `Error` is thrown if `canExit` is neither a function nor a boolean
         * @param {?} direction The direction in which this step should be left
         * @return {?} A [[Promise]] containing `true`, if the step can be exited in the given direction, false otherwise
         */
        function (direction) {
            return WizardStep.canTransitionStep(this.canExit, direction);
        };
        WizardStep.propDecorators = {
            stepTitleTemplate: [{ type: core.ContentChild, args: [WizardStepTitleDirective, { static: false },] }],
            stepSymbolTemplate: [{ type: core.ContentChild, args: [WizardStepSymbolDirective, { static: false },] }],
            stepId: [{ type: core.Input }],
            stepTitle: [{ type: core.Input }],
            navigationSymbol: [{ type: core.Input }],
            canEnter: [{ type: core.Input }],
            canExit: [{ type: core.Input }],
            stepEnter: [{ type: core.Output }],
            stepExit: [{ type: core.Output }],
            hidden: [{ type: core.HostBinding, args: ['hidden',] }]
        };
        return WizardStep;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Basic functionality every wizard completion step needs to provide
     *
     * @author Marc Arndt
     * @abstract
     */
    var   /**
     * Basic functionality every wizard completion step needs to provide
     *
     * @author Marc Arndt
     * @abstract
     */
    WizardCompletionStep = /** @class */ (function (_super) {
        __extends(WizardCompletionStep, _super);
        function WizardCompletionStep() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            /**
             * @inheritDoc
             */
            _this.stepExit = new core.EventEmitter();
            /**
             * @inheritDoc
             */
            _this.canExit = false;
            return _this;
        }
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @param {?} direction
         * @return {?}
         */
        WizardCompletionStep.prototype.enter = /**
         * @inheritDoc
         * @param {?} direction
         * @return {?}
         */
        function (direction) {
            this.completed = true;
            this.stepEnter.emit(direction);
        };
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @param {?} direction
         * @return {?}
         */
        WizardCompletionStep.prototype.exit = /**
         * @inheritDoc
         * @param {?} direction
         * @return {?}
         */
        function (direction) {
            // set this completion step as incomplete (unless it happens to be initiallyCompleted)
            this.completed = this.initiallyCompleted;
            this.stepExit.emit(direction);
        };
        return WizardCompletionStep;
    }(WizardStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `aw-wizard-completion-step` component can be used to define a completion/success step at the end of your wizard
     * After a `aw-wizard-completion-step` has been entered, it has the characteristic that the user is blocked from
     * leaving it again to a previous step.
     * In addition entering a `aw-wizard-completion-step` automatically sets the `aw-wizard` and all steps inside the `aw-wizard`
     * as completed.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-completion-step [stepTitle]="title of the wizard step"
     *    [navigationSymbol]="{ symbol: 'navigation symbol', fontFamily: 'navigation symbol font family' }"
     *    (stepEnter)="event emitter to be called when the wizard step is entered"
     *    (stepExit)="event emitter to be called when the wizard step is exited">
     *    ...
     * </aw-wizard-completion-step>
     * ```
     *
     * ### Example
     *
     * ```html
     * <aw-wizard-completion-step stepTitle="Step 1" [navigationSymbol]="{ symbol: '1' }">
     *    ...
     * </aw-wizard-completion-step>
     * ```
     *
     * With a navigation symbol from the `font-awesome` font:
     *
     * ```html
     * <aw-wizard-completion-step stepTitle="Step 1" [navigationSymbol]="{ symbol: '&#xf1ba;', fontFamily: 'FontAwesome' }">
     *    ...
     * </aw-wizard-completion-step>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardCompletionStepComponent = /** @class */ (function (_super) {
        __extends(WizardCompletionStepComponent, _super);
        function WizardCompletionStepComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WizardCompletionStepComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'aw-wizard-completion-step',
                        template: "<ng-content></ng-content>\n",
                        providers: [
                            { provide: WizardStep, useExisting: core.forwardRef((/**
                                 * @return {?}
                                 */
                                function () { return WizardCompletionStepComponent; })) },
                            { provide: WizardCompletionStep, useExisting: core.forwardRef((/**
                                 * @return {?}
                                 */
                                function () { return WizardCompletionStepComponent; })) }
                        ]
                    }] }
        ];
        return WizardCompletionStepComponent;
    }(WizardCompletionStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The direction in which a step transition was made
     *
     * @author Marc Arndt
     */
    /** @enum {number} */
    var MovingDirection = {
        /**
         * A forward step transition
         */
        Forwards: 0,
        /**
         * A backward step transition
         */
        Backwards: 1,
        /**
         * No step transition was done
         */
        Stay: 2,
    };
    MovingDirection[MovingDirection.Forwards] = 'Forwards';
    MovingDirection[MovingDirection.Backwards] = 'Backwards';
    MovingDirection[MovingDirection.Stay] = 'Stay';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Base implementation of [[NavigationMode]]
     *
     * Note: Built-in [[NavigationMode]] classes should be stateless, allowing the library user to easily create
     * an instance of a particular [[NavigationMode]] class and pass it to `<aw-wizard [navigationMode]="...">`.
     *
     * @author Marc Arndt
     * @abstract
     */
    var   /**
     * Base implementation of [[NavigationMode]]
     *
     * Note: Built-in [[NavigationMode]] classes should be stateless, allowing the library user to easily create
     * an instance of a particular [[NavigationMode]] class and pass it to `<aw-wizard [navigationMode]="...">`.
     *
     * @author Marc Arndt
     * @abstract
     */
    BaseNavigationMode = /** @class */ (function () {
        function BaseNavigationMode() {
        }
        /**
         * Checks, whether a wizard step, as defined by the given destination index, can be transitioned to.
         *
         * This method controls navigation by [[goToStep]], [[goToPreviousStep]], and [[goToNextStep]] directives.
         * Navigation by navigation bar is governed by [[isNavigable]].
         *
         * In this implementation, a destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - the destination step can be entered in the direction from the current step
         *
         * Subclasses can impose additional restrictions, see [[canTransitionToStep]].
         *
         * @param wizard The wizard component to operate on
         * @param destinationIndex The index of the destination step
         * @returns A [[Promise]] containing `true`, if the destination step can be transitioned to and `false` otherwise
         */
        /**
         * Checks, whether a wizard step, as defined by the given destination index, can be transitioned to.
         *
         * This method controls navigation by [[goToStep]], [[goToPreviousStep]], and [[goToNextStep]] directives.
         * Navigation by navigation bar is governed by [[isNavigable]].
         *
         * In this implementation, a destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - the destination step can be entered in the direction from the current step
         *
         * Subclasses can impose additional restrictions, see [[canTransitionToStep]].
         *
         * @param {?} wizard The wizard component to operate on
         * @param {?} destinationIndex The index of the destination step
         * @return {?} A [[Promise]] containing `true`, if the destination step can be transitioned to and `false` otherwise
         */
        BaseNavigationMode.prototype.canGoToStep = /**
         * Checks, whether a wizard step, as defined by the given destination index, can be transitioned to.
         *
         * This method controls navigation by [[goToStep]], [[goToPreviousStep]], and [[goToNextStep]] directives.
         * Navigation by navigation bar is governed by [[isNavigable]].
         *
         * In this implementation, a destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - the destination step can be entered in the direction from the current step
         *
         * Subclasses can impose additional restrictions, see [[canTransitionToStep]].
         *
         * @param {?} wizard The wizard component to operate on
         * @param {?} destinationIndex The index of the destination step
         * @return {?} A [[Promise]] containing `true`, if the destination step can be transitioned to and `false` otherwise
         */
        function (wizard, destinationIndex) {
            var _this = this;
            /** @type {?} */
            var hasStep = wizard.hasStep(destinationIndex);
            /** @type {?} */
            var movingDirection = wizard.getMovingDirection(destinationIndex);
            /** @type {?} */
            var canExitCurrentStep = (/**
             * @param {?} previous
             * @return {?}
             */
            function (previous) {
                return previous && wizard.currentStep.canExitStep(movingDirection);
            });
            /** @type {?} */
            var canEnterDestinationStep = (/**
             * @param {?} previous
             * @return {?}
             */
            function (previous) {
                return previous && wizard.getStepAtIndex(destinationIndex).canEnterStep(movingDirection);
            });
            /** @type {?} */
            var canTransitionToStep = (/**
             * @param {?} previous
             * @return {?}
             */
            function (previous) {
                return previous && _this.canTransitionToStep(wizard, destinationIndex);
            });
            return Promise.resolve(hasStep)
                .then(canTransitionToStep)
                // Apply user-defined checks at the end.  They can involve user interaction
                // which is better to be avoided if navigation mode does not actually allow the transition
                // (`canTransitionToStep` returns `false`).
                .then(canExitCurrentStep)
                .then(canEnterDestinationStep);
        };
        /**
         * Imposes additional restrictions for `canGoToStep` in current navigation mode.
         *
         * The base implementation allows transition iff the given step is navigable from the navigation bar (see `isNavigable`).
         * However, in some navigation modes `canTransitionToStep` can be more relaxed to allow navigation to certain steps
         * by previous/next buttons, but not using the navigation bar.
         *
         * @param wizard The wizard component to operate on
         * @param destinationIndex The index of the destination step
         * @returns `true`, if the destination step can be transitioned to and `false` otherwise
         */
        /**
         * Imposes additional restrictions for `canGoToStep` in current navigation mode.
         *
         * The base implementation allows transition iff the given step is navigable from the navigation bar (see `isNavigable`).
         * However, in some navigation modes `canTransitionToStep` can be more relaxed to allow navigation to certain steps
         * by previous/next buttons, but not using the navigation bar.
         *
         * @protected
         * @param {?} wizard The wizard component to operate on
         * @param {?} destinationIndex The index of the destination step
         * @return {?} `true`, if the destination step can be transitioned to and `false` otherwise
         */
        BaseNavigationMode.prototype.canTransitionToStep = /**
         * Imposes additional restrictions for `canGoToStep` in current navigation mode.
         *
         * The base implementation allows transition iff the given step is navigable from the navigation bar (see `isNavigable`).
         * However, in some navigation modes `canTransitionToStep` can be more relaxed to allow navigation to certain steps
         * by previous/next buttons, but not using the navigation bar.
         *
         * @protected
         * @param {?} wizard The wizard component to operate on
         * @param {?} destinationIndex The index of the destination step
         * @return {?} `true`, if the destination step can be transitioned to and `false` otherwise
         */
        function (wizard, destinationIndex) {
            return this.isNavigable(wizard, destinationIndex);
        };
        /**
         * Tries to transition to the wizard step, as denoted by the given destination index.
         *
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param wizard The wizard component to operate on
         * @param destinationIndex The index of the destination wizard step, which should be entered
         * @param preFinalize An event emitter, to be called before the step has been transitioned
         * @param postFinalize An event emitter, to be called after the step has been transitioned
         */
        /**
         * Tries to transition to the wizard step, as denoted by the given destination index.
         *
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param {?} wizard The wizard component to operate on
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        BaseNavigationMode.prototype.goToStep = /**
         * Tries to transition to the wizard step, as denoted by the given destination index.
         *
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param {?} wizard The wizard component to operate on
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        function (wizard, destinationIndex, preFinalize, postFinalize) {
            var _this = this;
            this.canGoToStep(wizard, destinationIndex).then((/**
             * @param {?} navigationAllowed
             * @return {?}
             */
            function (navigationAllowed) {
                if (navigationAllowed) {
                    // the current step can be exited in the given direction
                    /** @type {?} */
                    var movingDirection = wizard.getMovingDirection(destinationIndex);
                    /* istanbul ignore if */
                    if (preFinalize) {
                        preFinalize.emit();
                    }
                    // leave current step
                    wizard.currentStep.completed = true;
                    wizard.currentStep.exit(movingDirection);
                    wizard.currentStep.editing = false;
                    wizard.currentStep.selected = false;
                    _this.transition(wizard, destinationIndex);
                    // remember if the next step is already completed before entering it to properly set `editing` flag
                    /** @type {?} */
                    var wasCompleted = wizard.completed || wizard.currentStep.completed;
                    // go to next step
                    wizard.currentStep.enter(movingDirection);
                    wizard.currentStep.selected = true;
                    if (wasCompleted) {
                        wizard.currentStep.editing = true;
                    }
                    /* istanbul ignore if */
                    if (postFinalize) {
                        postFinalize.emit();
                    }
                }
                else {
                    // if the current step can't be left, reenter the current step
                    wizard.currentStep.exit(MovingDirection.Stay);
                    wizard.currentStep.enter(MovingDirection.Stay);
                }
            }));
        };
        /**
         * Transitions the wizard to the given step index.
         *
         * Can perform additional actions in particular navigation mode implementations.
         *
         * @param wizard The wizard component to operate on
         * @param destinationIndex The index of the destination wizard step
         */
        /**
         * Transitions the wizard to the given step index.
         *
         * Can perform additional actions in particular navigation mode implementations.
         *
         * @protected
         * @param {?} wizard The wizard component to operate on
         * @param {?} destinationIndex The index of the destination wizard step
         * @return {?}
         */
        BaseNavigationMode.prototype.transition = /**
         * Transitions the wizard to the given step index.
         *
         * Can perform additional actions in particular navigation mode implementations.
         *
         * @protected
         * @param {?} wizard The wizard component to operate on
         * @param {?} destinationIndex The index of the destination wizard step
         * @return {?}
         */
        function (wizard, destinationIndex) {
            wizard.currentStepIndex = destinationIndex;
        };
        /**
         * Resets the state of this wizard.
         *
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete.
         *
         * @param wizard The wizard component to operate on
         */
        /**
         * Resets the state of this wizard.
         *
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete.
         *
         * @param {?} wizard The wizard component to operate on
         * @return {?}
         */
        BaseNavigationMode.prototype.reset = /**
         * Resets the state of this wizard.
         *
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete.
         *
         * @param {?} wizard The wizard component to operate on
         * @return {?}
         */
        function (wizard) {
            this.ensureCanReset(wizard);
            // reset the step internal state
            wizard.wizardSteps.forEach((/**
             * @param {?} step
             * @return {?}
             */
            function (step) {
                step.completed = step.initiallyCompleted;
                step.selected = false;
                step.editing = false;
            }));
            // set the first step as the current step
            wizard.currentStepIndex = wizard.defaultStepIndex;
            wizard.currentStep.selected = true;
            wizard.currentStep.enter(MovingDirection.Forwards);
        };
        /**
         * Checks if wizard configuration allows to perform reset.
         *
         * A check failure is indicated by throwing an `Error` with the message discribing the discovered misconfiguration issue.
         *
         * Can include additional checks in particular navigation mode implementations.
         *
         * @param wizard The wizard component to operate on
         * @throws An `Error` is thrown, if a micconfiguration issue is discovered.
         */
        /**
         * Checks if wizard configuration allows to perform reset.
         *
         * A check failure is indicated by throwing an `Error` with the message discribing the discovered misconfiguration issue.
         *
         * Can include additional checks in particular navigation mode implementations.
         *
         * @throws An `Error` is thrown, if a micconfiguration issue is discovered.
         * @protected
         * @param {?} wizard The wizard component to operate on
         * @return {?}
         */
        BaseNavigationMode.prototype.ensureCanReset = /**
         * Checks if wizard configuration allows to perform reset.
         *
         * A check failure is indicated by throwing an `Error` with the message discribing the discovered misconfiguration issue.
         *
         * Can include additional checks in particular navigation mode implementations.
         *
         * @throws An `Error` is thrown, if a micconfiguration issue is discovered.
         * @protected
         * @param {?} wizard The wizard component to operate on
         * @return {?}
         */
        function (wizard) {
            // the wizard doesn't contain a step with the default step index
            if (!wizard.hasStep(wizard.defaultStepIndex)) {
                throw new Error("The wizard doesn't contain a step with index " + wizard.defaultStepIndex);
            }
        };
        return BaseNavigationMode;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The default navigation mode used by [[WizardComponent]] and [[NavigationModeDirective]].
     *
     * It is parameterized with two navigation policies passed to constructor:
     *
     * - [[navigateBackward]] policy controls whether wizard steps before the current step are navigable:
     *
     *   - `"deny"` -- the steps are not navigable
     *   - `"allow"` -- the steps are navigable
     *   - If the corresponding constructor argument is omitted or is `null` or `undefined`,
     *     then the default value is applied which is `"deny"`
     *
     * - [[navigateForward]] policy controls whether wizard steps after the current step are navigable:
     *
     *   - `"deny"` -- the steps are not navigable
     *   - `"allow"` -- the steps are navigable
     *   - `"visited"` -- a step is navigable iff it was already visited before
     *   - If the corresponding constructor argument is omitted or is `null` or `undefined`,
     *     then the default value is applied which is `"allow"`
     */
    var   /**
     * The default navigation mode used by [[WizardComponent]] and [[NavigationModeDirective]].
     *
     * It is parameterized with two navigation policies passed to constructor:
     *
     * - [[navigateBackward]] policy controls whether wizard steps before the current step are navigable:
     *
     *   - `"deny"` -- the steps are not navigable
     *   - `"allow"` -- the steps are navigable
     *   - If the corresponding constructor argument is omitted or is `null` or `undefined`,
     *     then the default value is applied which is `"deny"`
     *
     * - [[navigateForward]] policy controls whether wizard steps after the current step are navigable:
     *
     *   - `"deny"` -- the steps are not navigable
     *   - `"allow"` -- the steps are navigable
     *   - `"visited"` -- a step is navigable iff it was already visited before
     *   - If the corresponding constructor argument is omitted or is `null` or `undefined`,
     *     then the default value is applied which is `"allow"`
     */
    ConfigurableNavigationMode = /** @class */ (function (_super) {
        __extends(ConfigurableNavigationMode, _super);
        /**
         * Constructor
         *
         * @param navigateBackward Controls whether wizard steps before the current step are navigable
         * @param navigateForward Controls whether wizard steps before the current step are navigable
         */
        function ConfigurableNavigationMode(navigateBackward, navigateForward) {
            if (navigateBackward === void 0) { navigateBackward = null; }
            if (navigateForward === void 0) { navigateForward = null; }
            var _this = _super.call(this) || this;
            _this.navigateBackward = navigateBackward;
            _this.navigateForward = navigateForward;
            _this.navigateBackward = _this.navigateBackward || 'allow';
            _this.navigateForward = _this.navigateForward || 'deny';
            return _this;
        }
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @protected
         * @param {?} wizard
         * @param {?} destinationIndex
         * @return {?}
         */
        ConfigurableNavigationMode.prototype.canTransitionToStep = /**
         * @inheritDoc
         * @protected
         * @param {?} wizard
         * @param {?} destinationIndex
         * @return {?}
         */
        function (wizard, destinationIndex) {
            // if the destination step can be navigated to using the navigation bar,
            // it should be accessible with [goToStep] as well
            if (this.isNavigable(wizard, destinationIndex)) {
                return true;
            }
            // navigation with [goToStep] is permitted if all previous steps
            // to the destination step have been completed or are optional
            return wizard.wizardSteps
                .filter((/**
             * @param {?} step
             * @param {?} index
             * @return {?}
             */
            function (step, index) { return index < destinationIndex && index !== wizard.currentStepIndex; }))
                .every((/**
             * @param {?} step
             * @return {?}
             */
            function (step) { return step.completed || step.optional; }));
        };
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @protected
         * @param {?} wizard
         * @param {?} destinationIndex
         * @return {?}
         */
        ConfigurableNavigationMode.prototype.transition = /**
         * @inheritDoc
         * @protected
         * @param {?} wizard
         * @param {?} destinationIndex
         * @return {?}
         */
        function (wizard, destinationIndex) {
            if (this.navigateForward === 'deny') {
                // set all steps after the destination step to incomplete
                wizard.wizardSteps
                    .filter((/**
                 * @param {?} step
                 * @param {?} index
                 * @return {?}
                 */
                function (step, index) { return wizard.currentStepIndex > destinationIndex && index > destinationIndex; }))
                    .forEach((/**
                 * @param {?} step
                 * @return {?}
                 */
                function (step) { return step.completed = false; }));
            }
            _super.prototype.transition.call(this, wizard, destinationIndex);
        };
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @param {?} wizard
         * @param {?} destinationIndex
         * @return {?}
         */
        ConfigurableNavigationMode.prototype.isNavigable = /**
         * @inheritDoc
         * @param {?} wizard
         * @param {?} destinationIndex
         * @return {?}
         */
        function (wizard, destinationIndex) {
            // Check if the destination step can be navigated to
            /** @type {?} */
            var destinationStep = wizard.getStepAtIndex(destinationIndex);
            if (destinationStep instanceof WizardCompletionStep) {
                // A completion step can only be entered, if all previous steps have been completed, are optional, or selected
                /** @type {?} */
                var previousStepsCompleted = wizard.wizardSteps
                    .filter((/**
                 * @param {?} step
                 * @param {?} index
                 * @return {?}
                 */
                function (step, index) { return index < destinationIndex; }))
                    .every((/**
                 * @param {?} step
                 * @return {?}
                 */
                function (step) { return step.completed || step.optional || step.selected; }));
                if (!previousStepsCompleted) {
                    return false;
                }
            }
            // Apply navigation pocicies
            if (destinationIndex < wizard.currentStepIndex) {
                // If the destination step is before current, apply the `navigateBackward` policy
                switch (this.navigateBackward) {
                    case 'allow': return true;
                    case 'deny': return false;
                    default:
                        throw new Error("Invalid value for navigateBackward: " + this.navigateBackward);
                }
            }
            else if (destinationIndex > wizard.currentStepIndex) {
                // If the destination step is after current, apply the `navigateForward` policy
                switch (this.navigateForward) {
                    case 'allow': return true;
                    case 'deny': return false;
                    case 'visited': return destinationStep.completed;
                    default:
                        throw new Error("Invalid value for navigateForward: " + this.navigateForward);
                }
            }
            else {
                // Re-entering the current step is not allowed
                return false;
            }
        };
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @protected
         * @param {?} wizard
         * @return {?}
         */
        ConfigurableNavigationMode.prototype.ensureCanReset = /**
         * @inheritDoc
         * @protected
         * @param {?} wizard
         * @return {?}
         */
        function (wizard) {
            _super.prototype.ensureCanReset.call(this, wizard);
            // the default step is a completion step and the wizard contains more than one step
            /** @type {?} */
            var defaultWizardStep = wizard.getStepAtIndex(wizard.defaultStepIndex);
            /** @type {?} */
            var defaultCompletionStep = defaultWizardStep instanceof WizardCompletionStep;
            if (defaultCompletionStep && wizard.wizardSteps.length !== 1) {
                throw new Error("The default step index " + wizard.defaultStepIndex + " references a completion step");
            }
        };
        return ConfigurableNavigationMode;
    }(BaseNavigationMode));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `aw-wizard` component defines the root component of a wizard.
     * Through the setting of input parameters for the `aw-wizard` component it's possible to change the location and size
     * of its navigation bar.
     *
     * ### Syntax
     * ```html
     * <aw-wizard [navBarLocation]="location of navigation bar" [navBarLayout]="layout of navigation bar">
     *     ...
     * </aw-wizard>
     * ```
     *
     * ### Example
     *
     * Without completion step:
     *
     * ```html
     * <aw-wizard navBarLocation="top" navBarLayout="small">
     *     <aw-wizard-step>...</aw-wizard-step>
     *     <aw-wizard-step>...</aw-wizard-step>
     * </aw-wizard>
     * ```
     *
     * With completion step:
     *
     * ```html
     * <aw-wizard navBarLocation="top" navBarLayout="small">
     *     <aw-wizard-step>...</aw-wizard-step>
     *     <aw-wizard-step>...</aw-wizard-step>
     *     <aw-wizard-completion-step>...</aw-wizard-completion-step>
     * </aw-wizard>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardComponent = /** @class */ (function () {
        /**
         * Constructor
         */
        function WizardComponent() {
            /**
             * The location of the navigation bar inside the wizard.
             * This location can be either top, bottom, left or right
             */
            this.navBarLocation = 'top';
            /**
             * The layout of the navigation bar inside the wizard.
             * The layout can be either small, large-filled, large-empty or large-symbols
             */
            this.navBarLayout = 'small';
            /**
             * The direction in which the steps inside the navigation bar should be shown.
             * The direction can be either `left-to-right` or `right-to-left`
             */
            this.navBarDirection = 'left-to-right';
            this._defaultStepIndex = 0;
            /**
             * True, if the navigation bar shouldn't be used for navigating
             */
            this.disableNavigationBar = false;
            /**
             * The navigation mode used to navigate inside the wizard
             *
             * For outside access, use the [[navigation]] getter.
             */
            this._navigation = new ConfigurableNavigationMode();
            /**
             * An array representation of all wizard steps belonging to this model
             *
             * For outside access, use the [[wizardSteps]] getter.
             */
            this._wizardSteps = [];
            /**
             * The index of the currently visible and selected step inside the wizardSteps QueryList.
             * If this wizard contains no steps, currentStepIndex is -1
             *
             * Note: Do not modify this field directly.  Instead, use navigation methods:
             * [[goToStep]], [[goToPreviousStep]], [[goToNextStep]].
             */
            this.currentStepIndex = -1;
        }
        Object.defineProperty(WizardComponent.prototype, "defaultStepIndex", {
            /**
             * The initially selected step, represented by its index
             * Beware: This initial default is only used if no wizard step has been enhanced with the `selected` directive
             */
            get: /**
             * The initially selected step, represented by its index
             * Beware: This initial default is only used if no wizard step has been enhanced with the `selected` directive
             * @return {?}
             */
            function () {
                // This value can be either:
                // - the index of a wizard step with a `selected` directive, or
                // - the default step index, set in the [[WizardComponent]]
                // This value can be either:
                // - the index of a wizard step with a `selected` directive, or
                // - the default step index, set in the [[WizardComponent]]
                /** @type {?} */
                var foundDefaultStep = this.wizardSteps.find((/**
                 * @param {?} step
                 * @return {?}
                 */
                function (step) { return step.defaultSelected; }));
                if (foundDefaultStep) {
                    return this.getIndexOfStep(foundDefaultStep);
                }
                else {
                    return this._defaultStepIndex;
                }
            },
            set: /**
             * @param {?} defaultStepIndex
             * @return {?}
             */
            function (defaultStepIndex) {
                this._defaultStepIndex = defaultStepIndex;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardComponent.prototype, "horizontalOrientation", {
            /**
             * Returns true if this wizard uses a horizontal orientation.
             * The wizard uses a horizontal orientation, iff the navigation bar is shown at the top or bottom of this wizard
             *
             * @returns True if this wizard uses a horizontal orientation
             */
            get: /**
             * Returns true if this wizard uses a horizontal orientation.
             * The wizard uses a horizontal orientation, iff the navigation bar is shown at the top or bottom of this wizard
             *
             * @return {?} True if this wizard uses a horizontal orientation
             */
            function () {
                return this.navBarLocation === 'top' || this.navBarLocation === 'bottom';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardComponent.prototype, "verticalOrientation", {
            /**
             * Returns true if this wizard uses a vertical orientation.
             * The wizard uses a vertical orientation, iff the navigation bar is shown at the left or right of this wizard
             *
             * @returns True if this wizard uses a vertical orientation
             */
            get: /**
             * Returns true if this wizard uses a vertical orientation.
             * The wizard uses a vertical orientation, iff the navigation bar is shown at the left or right of this wizard
             *
             * @return {?} True if this wizard uses a vertical orientation
             */
            function () {
                return this.navBarLocation === 'left' || this.navBarLocation === 'right';
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        WizardComponent.prototype.ngAfterContentInit = /**
         * Initialization work
         * @return {?}
         */
        function () {
            var _this = this;
            // add a subscriber to the wizard steps QueryList to listen to changes in the DOM
            this.wizardStepsQueryList.changes.subscribe((/**
             * @param {?} changedWizardSteps
             * @return {?}
             */
            function (changedWizardSteps) {
                _this.updateWizardSteps(changedWizardSteps.toArray());
            }));
            // initialize the model
            this.updateWizardSteps(this.wizardStepsQueryList.toArray());
            // finally reset the whole wizard componennt
            this.reset();
        };
        Object.defineProperty(WizardComponent.prototype, "currentStep", {
            /**
             * The WizardStep object belonging to the currently visible and selected step.
             * The currentStep is always the currently selected wizard step.
             * The currentStep can be either completed, if it was visited earlier,
             * or not completed, if it is visited for the first time or its state is currently out of date.
             *
             * If this wizard contains no steps, currentStep is null
             */
            get: /**
             * The WizardStep object belonging to the currently visible and selected step.
             * The currentStep is always the currently selected wizard step.
             * The currentStep can be either completed, if it was visited earlier,
             * or not completed, if it is visited for the first time or its state is currently out of date.
             *
             * If this wizard contains no steps, currentStep is null
             * @return {?}
             */
            function () {
                if (this.hasStep(this.currentStepIndex)) {
                    return this.wizardSteps[this.currentStepIndex];
                }
                else {
                    return null;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardComponent.prototype, "completed", {
            /**
             * The completeness of the wizard.
             * If the wizard has been completed, i.e. all steps are either completed or optional, this value is true, otherwise it is false
             */
            get: /**
             * The completeness of the wizard.
             * If the wizard has been completed, i.e. all steps are either completed or optional, this value is true, otherwise it is false
             * @return {?}
             */
            function () {
                return this.wizardSteps.every((/**
                 * @param {?} step
                 * @return {?}
                 */
                function (step) { return step.completed || step.optional; }));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardComponent.prototype, "wizardSteps", {
            /**
             * An array representation of all wizard steps belonging to this model
             */
            get: /**
             * An array representation of all wizard steps belonging to this model
             * @return {?}
             */
            function () {
                return this._wizardSteps;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Updates the wizard steps to the new array
         *
         * @param wizardSteps The updated wizard steps
         */
        /**
         * Updates the wizard steps to the new array
         *
         * @private
         * @param {?} wizardSteps The updated wizard steps
         * @return {?}
         */
        WizardComponent.prototype.updateWizardSteps = /**
         * Updates the wizard steps to the new array
         *
         * @private
         * @param {?} wizardSteps The updated wizard steps
         * @return {?}
         */
        function (wizardSteps) {
            // the wizard is currently not in the initialization phase
            if (this.wizardSteps.length > 0 && this.currentStepIndex > -1) {
                this.currentStepIndex = wizardSteps.indexOf(this.wizardSteps[this.currentStepIndex]);
            }
            this._wizardSteps = wizardSteps;
        };
        Object.defineProperty(WizardComponent.prototype, "navigation", {
            /**
             * The navigation mode used to navigate inside the wizard
             */
            get: /**
             * The navigation mode used to navigate inside the wizard
             * @return {?}
             */
            function () {
                return this._navigation;
            },
            /**
             * Updates the navigation mode for this wizard component
             *
             * @param navigation The updated navigation mode
             */
            set: /**
             * Updates the navigation mode for this wizard component
             *
             * @param {?} navigation The updated navigation mode
             * @return {?}
             */
            function (navigation) {
                this._navigation = navigation;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Checks if a given index `stepIndex` is inside the range of possible wizard steps inside this wizard
         *
         * @param stepIndex The to be checked index of a step inside this wizard
         * @returns True if the given `stepIndex` is contained inside this wizard, false otherwise
         */
        /**
         * Checks if a given index `stepIndex` is inside the range of possible wizard steps inside this wizard
         *
         * @param {?} stepIndex The to be checked index of a step inside this wizard
         * @return {?} True if the given `stepIndex` is contained inside this wizard, false otherwise
         */
        WizardComponent.prototype.hasStep = /**
         * Checks if a given index `stepIndex` is inside the range of possible wizard steps inside this wizard
         *
         * @param {?} stepIndex The to be checked index of a step inside this wizard
         * @return {?} True if the given `stepIndex` is contained inside this wizard, false otherwise
         */
        function (stepIndex) {
            return this.wizardSteps.length > 0 && 0 <= stepIndex && stepIndex < this.wizardSteps.length;
        };
        /**
         * Checks if this wizard has a previous step, compared to the current step
         *
         * @returns True if this wizard has a previous step before the current step
         */
        /**
         * Checks if this wizard has a previous step, compared to the current step
         *
         * @return {?} True if this wizard has a previous step before the current step
         */
        WizardComponent.prototype.hasPreviousStep = /**
         * Checks if this wizard has a previous step, compared to the current step
         *
         * @return {?} True if this wizard has a previous step before the current step
         */
        function () {
            return this.hasStep(this.currentStepIndex - 1);
        };
        /**
         * Checks if this wizard has a next step, compared to the current step
         *
         * @returns True if this wizard has a next step after the current step
         */
        /**
         * Checks if this wizard has a next step, compared to the current step
         *
         * @return {?} True if this wizard has a next step after the current step
         */
        WizardComponent.prototype.hasNextStep = /**
         * Checks if this wizard has a next step, compared to the current step
         *
         * @return {?} True if this wizard has a next step after the current step
         */
        function () {
            return this.hasStep(this.currentStepIndex + 1);
        };
        /**
         * Checks if this wizard is currently inside its last step
         *
         * @returns True if the wizard is currently inside its last step
         */
        /**
         * Checks if this wizard is currently inside its last step
         *
         * @return {?} True if the wizard is currently inside its last step
         */
        WizardComponent.prototype.isLastStep = /**
         * Checks if this wizard is currently inside its last step
         *
         * @return {?} True if the wizard is currently inside its last step
         */
        function () {
            return this.wizardSteps.length > 0 && this.currentStepIndex === this.wizardSteps.length - 1;
        };
        /**
         * Finds the [[WizardStep]] at the given index `stepIndex`.
         * If no [[WizardStep]] exists at the given index an Error is thrown
         *
         * @param stepIndex The given index
         * @returns The found [[WizardStep]] at the given index `stepIndex`
         * @throws An `Error` is thrown, if the given index `stepIndex` doesn't exist
         */
        /**
         * Finds the [[WizardStep]] at the given index `stepIndex`.
         * If no [[WizardStep]] exists at the given index an Error is thrown
         *
         * @throws An `Error` is thrown, if the given index `stepIndex` doesn't exist
         * @param {?} stepIndex The given index
         * @return {?} The found [[WizardStep]] at the given index `stepIndex`
         */
        WizardComponent.prototype.getStepAtIndex = /**
         * Finds the [[WizardStep]] at the given index `stepIndex`.
         * If no [[WizardStep]] exists at the given index an Error is thrown
         *
         * @throws An `Error` is thrown, if the given index `stepIndex` doesn't exist
         * @param {?} stepIndex The given index
         * @return {?} The found [[WizardStep]] at the given index `stepIndex`
         */
        function (stepIndex) {
            if (!this.hasStep(stepIndex)) {
                throw new Error("Expected a known step, but got stepIndex: " + stepIndex + ".");
            }
            return this.wizardSteps[stepIndex];
        };
        /**
         * Finds the index of the step with the given `stepId`.
         * If no step with the given `stepId` exists, `-1` is returned
         *
         * @param stepId The given step id
         * @returns The found index of a step with the given step id, or `-1` if no step with the given id is included in the wizard
         */
        /**
         * Finds the index of the step with the given `stepId`.
         * If no step with the given `stepId` exists, `-1` is returned
         *
         * @param {?} stepId The given step id
         * @return {?} The found index of a step with the given step id, or `-1` if no step with the given id is included in the wizard
         */
        WizardComponent.prototype.getIndexOfStepWithId = /**
         * Finds the index of the step with the given `stepId`.
         * If no step with the given `stepId` exists, `-1` is returned
         *
         * @param {?} stepId The given step id
         * @return {?} The found index of a step with the given step id, or `-1` if no step with the given id is included in the wizard
         */
        function (stepId) {
            return this.wizardSteps.findIndex((/**
             * @param {?} step
             * @return {?}
             */
            function (step) { return step.stepId === stepId; }));
        };
        /**
         * Finds the index of the given [[WizardStep]] `step`.
         * If the given [[WizardStep]] is not contained inside this wizard, `-1` is returned
         *
         * @param step The given [[WizardStep]]
         * @returns The found index of `step` or `-1` if the step is not included in the wizard
         */
        /**
         * Finds the index of the given [[WizardStep]] `step`.
         * If the given [[WizardStep]] is not contained inside this wizard, `-1` is returned
         *
         * @param {?} step The given [[WizardStep]]
         * @return {?} The found index of `step` or `-1` if the step is not included in the wizard
         */
        WizardComponent.prototype.getIndexOfStep = /**
         * Finds the index of the given [[WizardStep]] `step`.
         * If the given [[WizardStep]] is not contained inside this wizard, `-1` is returned
         *
         * @param {?} step The given [[WizardStep]]
         * @return {?} The found index of `step` or `-1` if the step is not included in the wizard
         */
        function (step) {
            return this.wizardSteps.indexOf(step);
        };
        /**
         * Calculates the correct [[MovingDirection]] value for a given `destinationStep` compared to the `currentStepIndex`.
         *
         * @param destinationStep The given destination step
         * @returns The calculated [[MovingDirection]]
         */
        /**
         * Calculates the correct [[MovingDirection]] value for a given `destinationStep` compared to the `currentStepIndex`.
         *
         * @param {?} destinationStep The given destination step
         * @return {?} The calculated [[MovingDirection]]
         */
        WizardComponent.prototype.getMovingDirection = /**
         * Calculates the correct [[MovingDirection]] value for a given `destinationStep` compared to the `currentStepIndex`.
         *
         * @param {?} destinationStep The given destination step
         * @return {?} The calculated [[MovingDirection]]
         */
        function (destinationStep) {
            /** @type {?} */
            var movingDirection;
            if (destinationStep > this.currentStepIndex) {
                movingDirection = MovingDirection.Forwards;
            }
            else if (destinationStep < this.currentStepIndex) {
                movingDirection = MovingDirection.Backwards;
            }
            else {
                movingDirection = MovingDirection.Stay;
            }
            return movingDirection;
        };
        /**
         * Checks, whether a wizard step, as defined by the given destination index, can be transitioned to.
         *
         * This method controls navigation by [[goToStep]], [[goToPreviousStep]], and [[goToNextStep]] directives.
         * Navigation by navigation bar is governed by [[isNavigable]].
         *
         * @param destinationIndex The index of the destination step
         * @returns A [[Promise]] containing `true`, if the destination step can be transitioned to and false otherwise
         */
        /**
         * Checks, whether a wizard step, as defined by the given destination index, can be transitioned to.
         *
         * This method controls navigation by [[goToStep]], [[goToPreviousStep]], and [[goToNextStep]] directives.
         * Navigation by navigation bar is governed by [[isNavigable]].
         *
         * @param {?} destinationIndex The index of the destination step
         * @return {?} A [[Promise]] containing `true`, if the destination step can be transitioned to and false otherwise
         */
        WizardComponent.prototype.canGoToStep = /**
         * Checks, whether a wizard step, as defined by the given destination index, can be transitioned to.
         *
         * This method controls navigation by [[goToStep]], [[goToPreviousStep]], and [[goToNextStep]] directives.
         * Navigation by navigation bar is governed by [[isNavigable]].
         *
         * @param {?} destinationIndex The index of the destination step
         * @return {?} A [[Promise]] containing `true`, if the destination step can be transitioned to and false otherwise
         */
        function (destinationIndex) {
            return this.navigation.canGoToStep(this, destinationIndex);
        };
        /**
         * Tries to transition to the wizard step, as denoted by the given destination index.
         *
         * Note: You do not have to call [[canGoToStep]] before calling [[goToStep]].
         * The [[canGoToStep]] method will be called automatically.
         *
         * @param destinationIndex The index of the destination wizard step, which should be entered
         * @param preFinalize An event emitter, to be called before the step has been transitioned
         * @param postFinalize An event emitter, to be called after the step has been transitioned
         */
        /**
         * Tries to transition to the wizard step, as denoted by the given destination index.
         *
         * Note: You do not have to call [[canGoToStep]] before calling [[goToStep]].
         * The [[canGoToStep]] method will be called automatically.
         *
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        WizardComponent.prototype.goToStep = /**
         * Tries to transition to the wizard step, as denoted by the given destination index.
         *
         * Note: You do not have to call [[canGoToStep]] before calling [[goToStep]].
         * The [[canGoToStep]] method will be called automatically.
         *
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        function (destinationIndex, preFinalize, postFinalize) {
            return this.navigation.goToStep(this, destinationIndex, preFinalize, postFinalize);
        };
        /**
         * Tries to transition the wizard to the previous step
         *
         * @param preFinalize An event emitter, to be called before the step has been transitioned
         * @param postFinalize An event emitter, to be called after the step has been transitioned
         */
        /**
         * Tries to transition the wizard to the previous step
         *
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        WizardComponent.prototype.goToPreviousStep = /**
         * Tries to transition the wizard to the previous step
         *
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        function (preFinalize, postFinalize) {
            return this.navigation.goToStep(this, this.currentStepIndex - 1, preFinalize, postFinalize);
        };
        /**
         * Tries to transition the wizard to the next step
         *
         * @param preFinalize An event emitter, to be called before the step has been transitioned
         * @param postFinalize An event emitter, to be called after the step has been transitioned
         */
        /**
         * Tries to transition the wizard to the next step
         *
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        WizardComponent.prototype.goToNextStep = /**
         * Tries to transition the wizard to the next step
         *
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        function (preFinalize, postFinalize) {
            return this.navigation.goToStep(this, this.currentStepIndex + 1, preFinalize, postFinalize);
        };
        /**
         * Checks, whether the wizard step, located at the given index, can be navigated to using the navigation bar.
         *
         * @param destinationIndex The index of the destination step
         * @returns True if the step can be navigated to, false otherwise
         */
        /**
         * Checks, whether the wizard step, located at the given index, can be navigated to using the navigation bar.
         *
         * @param {?} destinationIndex The index of the destination step
         * @return {?} True if the step can be navigated to, false otherwise
         */
        WizardComponent.prototype.isNavigable = /**
         * Checks, whether the wizard step, located at the given index, can be navigated to using the navigation bar.
         *
         * @param {?} destinationIndex The index of the destination step
         * @return {?} True if the step can be navigated to, false otherwise
         */
        function (destinationIndex) {
            return this.navigation.isNavigable(this, destinationIndex);
        };
        /**
         * Resets the state of this wizard.
         */
        /**
         * Resets the state of this wizard.
         * @return {?}
         */
        WizardComponent.prototype.reset = /**
         * Resets the state of this wizard.
         * @return {?}
         */
        function () {
            this.navigation.reset(this);
        };
        WizardComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'aw-wizard',
                        template: "<aw-wizard-navigation-bar\n  [direction]=\"navBarDirection\"\n  *ngIf=\"navBarLocation == 'top' || navBarLocation == 'left'\"\n  [ngClass]=\"{\n    vertical: navBarLocation == 'left',\n    horizontal: navBarLocation == 'top',\n    small: navBarLayout == 'small',\n    'large-filled': navBarLayout == 'large-filled',\n    'large-filled-symbols': navBarLayout == 'large-filled-symbols',\n    'large-empty': navBarLayout == 'large-empty',\n    'large-empty-symbols': navBarLayout == 'large-empty-symbols'\n  }\">\n</aw-wizard-navigation-bar>\n\n<div [ngClass]=\"{\n  'wizard-steps': true,\n  vertical: navBarLocation == 'left' || navBarLocation == 'right',\n  horizontal: navBarLocation == 'top' || navBarLocation == 'bottom'\n}\">\n  <ng-content></ng-content>\n</div>\n\n<aw-wizard-navigation-bar\n  [direction]=\"navBarDirection\"\n  *ngIf=\"navBarLocation == 'bottom' || navBarLocation == 'right'\"\n  [ngClass]=\"{\n    vertical: navBarLocation == 'right',\n    horizontal: navBarLocation == 'bottom',\n    small: navBarLayout == 'small',\n    'large-filled': navBarLayout == 'large-filled',\n    'large-filled-symbols': navBarLayout == 'large-filled-symbols',\n    'large-empty': navBarLayout == 'large-empty',\n    'large-empty-symbols': navBarLayout == 'large-empty-symbols'\n  }\">\n</aw-wizard-navigation-bar>\n"
                    }] }
        ];
        /** @nocollapse */
        WizardComponent.ctorParameters = function () { return []; };
        WizardComponent.propDecorators = {
            wizardStepsQueryList: [{ type: core.ContentChildren, args: [WizardStep,] }],
            navBarLocation: [{ type: core.Input }],
            navBarLayout: [{ type: core.Input }],
            navBarDirection: [{ type: core.Input }],
            defaultStepIndex: [{ type: core.Input }],
            disableNavigationBar: [{ type: core.Input }],
            horizontalOrientation: [{ type: core.HostBinding, args: ['class.horizontal',] }],
            verticalOrientation: [{ type: core.HostBinding, args: ['class.vertical',] }]
        };
        return WizardComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `aw-wizard-navigation-bar` component contains the navigation bar inside a [[WizardComponent]].
     * To correctly display the navigation bar, it's required to set the right css classes for the navigation bar,
     * otherwise it will look like a normal `ul` component.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-navigation-bar></aw-wizard-navigation-bar>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardNavigationBarComponent = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizard The state the wizard currently resides in
         */
        function WizardNavigationBarComponent(wizard) {
            this.wizard = wizard;
            /**
             * The direction in which the wizard steps should be shown in the navigation bar.
             * This value can be either `left-to-right` or `right-to-left`
             */
            this.direction = 'left-to-right';
        }
        Object.defineProperty(WizardNavigationBarComponent.prototype, "wizardSteps", {
            /**
             * Returns all [[WizardStep]]s contained in the wizard
             *
             * @returns An array containing all [[WizardStep]]s
             */
            get: /**
             * Returns all [[WizardStep]]s contained in the wizard
             *
             * @return {?} An array containing all [[WizardStep]]s
             */
            function () {
                switch (this.direction) {
                    case 'right-to-left':
                        return this.wizard.wizardSteps.slice().reverse();
                    case 'left-to-right':
                    default:
                        return this.wizard.wizardSteps;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardNavigationBarComponent.prototype, "numberOfWizardSteps", {
            /**
             * Returns the number of wizard steps, that need to be displaced in the navigation bar
             *
             * @returns The number of wizard steps to be displayed
             */
            get: /**
             * Returns the number of wizard steps, that need to be displaced in the navigation bar
             *
             * @return {?} The number of wizard steps to be displayed
             */
            function () {
                return this.wizard.wizardSteps.length;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Checks, whether a [[WizardStep]] can be marked as `current` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as `current`
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `current` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `current`
         */
        WizardNavigationBarComponent.prototype.isCurrent = /**
         * Checks, whether a [[WizardStep]] can be marked as `current` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `current`
         */
        function (wizardStep) {
            return wizardStep.selected;
        };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `editing` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as `editing`
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `editing` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `editing`
         */
        WizardNavigationBarComponent.prototype.isEditing = /**
         * Checks, whether a [[WizardStep]] can be marked as `editing` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `editing`
         */
        function (wizardStep) {
            return wizardStep.editing;
        };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `done` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as `done`
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `done` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `done`
         */
        WizardNavigationBarComponent.prototype.isDone = /**
         * Checks, whether a [[WizardStep]] can be marked as `done` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `done`
         */
        function (wizardStep) {
            return wizardStep.completed;
        };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `optional` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as `optional`
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `optional` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `optional`
         */
        WizardNavigationBarComponent.prototype.isOptional = /**
         * Checks, whether a [[WizardStep]] can be marked as `optional` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `optional`
         */
        function (wizardStep) {
            return wizardStep.optional;
        };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `completed` in the navigation bar.
         *
         * The `completed` class is only applied to completion steps.
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as `completed`
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `completed` in the navigation bar.
         *
         * The `completed` class is only applied to completion steps.
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `completed`
         */
        WizardNavigationBarComponent.prototype.isCompleted = /**
         * Checks, whether a [[WizardStep]] can be marked as `completed` in the navigation bar.
         *
         * The `completed` class is only applied to completion steps.
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as `completed`
         */
        function (wizardStep) {
            return wizardStep instanceof WizardCompletionStep && this.wizard.completed;
        };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `navigable` in the navigation bar.
         * A wizard step can be navigated to if:
         * - the step is currently not selected
         * - the navigation bar isn't disabled
         * - the navigation mode allows navigation to the step
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as navigable
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `navigable` in the navigation bar.
         * A wizard step can be navigated to if:
         * - the step is currently not selected
         * - the navigation bar isn't disabled
         * - the navigation mode allows navigation to the step
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as navigable
         */
        WizardNavigationBarComponent.prototype.isNavigable = /**
         * Checks, whether a [[WizardStep]] can be marked as `navigable` in the navigation bar.
         * A wizard step can be navigated to if:
         * - the step is currently not selected
         * - the navigation bar isn't disabled
         * - the navigation mode allows navigation to the step
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as navigable
         */
        function (wizardStep) {
            return !wizardStep.selected && !this.wizard.disableNavigationBar &&
                this.wizard.isNavigable(this.wizard.getIndexOfStep(wizardStep));
        };
        WizardNavigationBarComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'aw-wizard-navigation-bar',
                        template: "<ul class=\"steps-indicator steps-{{numberOfWizardSteps}}\">\n  <li [attr.id]=\"step.stepId\" *ngFor=\"let step of wizardSteps\"\n      [ngClass]=\"{\n        current: isCurrent(step),\n        editing: isEditing(step),\n        done: isDone(step),\n        optional: isOptional(step),\n        completed: isCompleted(step),\n        navigable: isNavigable(step)\n  }\">\n    <a [awGoToStep]=\"step\">\n      <div class=\"label\">\n        <ng-container *ngIf=\"step.stepTitleTemplate\" [ngTemplateOutlet]=\"step.stepTitleTemplate.templateRef\"></ng-container>\n        <ng-container *ngIf=\"!step.stepTitleTemplate\">{{step.stepTitle}}</ng-container>\n      </div>\n      <div class=\"step-indicator\" [ngStyle]=\"{ 'font-family': step.stepSymbolTemplate ? '' : step.navigationSymbol.fontFamily }\">\n        <ng-container *ngIf=\"step.stepSymbolTemplate\" [ngTemplateOutlet]=\"step.stepSymbolTemplate.templateRef\"></ng-container>\n        <ng-container *ngIf=\"!step.stepSymbolTemplate\">{{step.navigationSymbol.symbol}}</ng-container>\n      </div>\n    </a>\n  </li>\n</ul>\n"
                    }] }
        ];
        /** @nocollapse */
        WizardNavigationBarComponent.ctorParameters = function () { return [
            { type: WizardComponent }
        ]; };
        WizardNavigationBarComponent.propDecorators = {
            direction: [{ type: core.Input }]
        };
        return WizardNavigationBarComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `aw-wizard-step` component is used to define a normal step inside a wizard.
     *
     * ### Syntax
     *
     * With `stepTitle` and `navigationSymbol` inputs:
     *
     * ```html
     * <aw-wizard-step [stepTitle]="step title" [navigationSymbol]="{ symbol: 'symbol', fontFamily: 'font-family' }"
     *    [canExit]="deciding function" (stepEnter)="enter function" (stepExit)="exit function">
     *    ...
     * </aw-wizard-step>
     * ```
     *
     * With `awWizardStepTitle` and `awWizardStepSymbol` directives:
     *
     * ```html
     * <aw-wizard-step"
     *    [canExit]="deciding function" (stepEnter)="enter function" (stepExit)="exit function">
     *    <ng-template awWizardStepTitle>
     *        step title
     *    </ng-template>
     *    <ng-template awWizardStepSymbol>
     *        symbol
     *    </ng-template>
     *    ...
     * </aw-wizard-step>
     * ```
     *
     * ### Example
     *
     * With `stepTitle` and `navigationSymbol` inputs:
     *
     * ```html
     * <aw-wizard-step stepTitle="Address information" [navigationSymbol]="{ symbol: '&#xf1ba;', fontFamily: 'FontAwesome' }">
     *    ...
     * </aw-wizard-step>
     * ```
     *
     * With `awWizardStepTitle` and `awWizardStepSymbol` directives:
     *
     * ```html
     * <aw-wizard-step>
     *    <ng-template awWizardStepTitle>
     *        Address information
     *    </ng-template>
     *    <ng-template awWizardStepSymbol>
     *        <i class="fa fa-taxi"></i>
     *    </ng-template>
     * </aw-wizard-step>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardStepComponent = /** @class */ (function (_super) {
        __extends(WizardStepComponent, _super);
        function WizardStepComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WizardStepComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'aw-wizard-step',
                        template: "<ng-content></ng-content>\n",
                        providers: [
                            { provide: WizardStep, useExisting: core.forwardRef((/**
                                 * @return {?}
                                 */
                                function () { return WizardStepComponent; })) }
                        ]
                    }] }
        ];
        return WizardStepComponent;
    }(WizardStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awEnableBackLinks` directive can be used to allow the user to leave a [[WizardCompletionStep]] after is has been entered.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-completion-step awEnableBackLinks (stepExit)="exit function">
     *     ...
     * </aw-wizard-completion-step>
     * ```
     *
     * ### Example
     *
     * ```html
     * <aw-wizard-completion-step stepTitle="Final step" awEnableBackLinks>
     *     ...
     * </aw-wizard-completion-step>
     * ```
     *
     * @author Marc Arndt
     */
    var EnableBackLinksDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param completionStep The wizard completion step, which should be exitable
         */
        function EnableBackLinksDirective(completionStep) {
            this.completionStep = completionStep;
            /**
             * This EventEmitter is called when the step is exited.
             * The bound method can be used to do cleanup work.
             */
            this.stepExit = new core.EventEmitter();
        }
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        EnableBackLinksDirective.prototype.ngOnInit = /**
         * Initialization work
         * @return {?}
         */
        function () {
            this.completionStep.canExit = true;
            this.completionStep.stepExit = this.stepExit;
        };
        EnableBackLinksDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awEnableBackLinks]'
                    },] }
        ];
        /** @nocollapse */
        EnableBackLinksDirective.ctorParameters = function () { return [
            { type: WizardCompletionStep, decorators: [{ type: core.Host }] }
        ]; };
        EnableBackLinksDirective.propDecorators = {
            stepExit: [{ type: core.Output }]
        };
        return EnableBackLinksDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Checks whether the given `value` implements the interface [[StepId]].
     *
     * @param {?} value The value to be checked
     * @return {?} True if the given value implements [[StepId]] and false otherwise
     */
    function isStepId(value) {
        return value.hasOwnProperty('stepId') && !(value instanceof WizardStep);
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Checks whether the given `value` implements the interface [[StepIndex]].
     *
     * @param {?} value The value to be checked
     * @return {?} True if the given value implements [[StepIndex]] and false otherwise
     */
    function isStepIndex(value) {
        return value.hasOwnProperty('stepIndex');
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Checks whether the given `value` implements the interface [[StepOffset]].
     *
     * @param {?} value The value to be checked
     * @return {?} True if the given value implements [[StepOffset]] and false otherwise
     */
    function isStepOffset(value) {
        return value.hasOwnProperty('stepOffset');
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awGoToStep` directive can be used to navigate to a given step.
     * This step can be defined in one of multiple formats
     *
     * ### Syntax
     *
     * With absolute step index:
     *
     * ```html
     * <button [awGoToStep]="{ stepIndex: absolute step index }" (finalize)="finalize method">...</button>
     * ```
     *
     * With unique step id:
     *
     * ```html
     * <button [awGoToStep]="{ stepId: 'step id of destination step' }" (finalize)="finalize method">...</button>
     * ```
     *
     * With a wizard step object:
     *
     * ```html
     * <button [awGoToStep]="wizard step object" (finalize)="finalize method">...</button>
     * ```
     *
     * With an offset to the defining step:
     *
     * ```html
     * <button [awGoToStep]="{ stepOffset: offset }" (finalize)="finalize method">...</button>
     * ```
     *
     * @author Marc Arndt
     */
    var GoToStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizard The wizard component
         * @param wizardStep The wizard step, which contains this [[GoToStepDirective]]
         */
        function GoToStepDirective(wizard, wizardStep) {
            this.wizard = wizard;
            this.wizardStep = wizardStep;
            /**
             * This [[EventEmitter]] is called directly before the current step is exited during a transition through a component with this directive.
             */
            this.preFinalize = new core.EventEmitter();
            /**
             * This [[EventEmitter]] is called directly after the current step is exited during a transition through a component with this directive.
             */
            this.postFinalize = new core.EventEmitter();
        }
        Object.defineProperty(GoToStepDirective.prototype, "finalize", {
            /**
             * A convenience field for `preFinalize`
             */
            get: /**
             * A convenience field for `preFinalize`
             * @return {?}
             */
            function () {
                return this.preFinalize;
            },
            /**
             * A convenience name for `preFinalize`
             *
             * @param emitter The [[EventEmitter]] to be set
             */
            set: /**
             * A convenience name for `preFinalize`
             *
             * @param {?} emitter The [[EventEmitter]] to be set
             * @return {?}
             */
            function (emitter) {
                /* istanbul ignore next */
                this.preFinalize = emitter;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(GoToStepDirective.prototype, "destinationStep", {
            /**
             * Returns the destination step of this directive as an absolute step index inside the wizard
             *
             * @returns The index of the destination step
             * @throws If `targetStep` is of an unknown type an `Error` is thrown
             */
            get: /**
             * Returns the destination step of this directive as an absolute step index inside the wizard
             *
             * @throws If `targetStep` is of an unknown type an `Error` is thrown
             * @return {?} The index of the destination step
             */
            function () {
                /** @type {?} */
                var destinationStep;
                if (isStepIndex(this.targetStep)) {
                    destinationStep = this.targetStep.stepIndex;
                }
                else if (isStepId(this.targetStep)) {
                    destinationStep = this.wizard.getIndexOfStepWithId(this.targetStep.stepId);
                }
                else if (isStepOffset(this.targetStep) && this.wizardStep !== null) {
                    destinationStep = this.wizard.getIndexOfStep(this.wizardStep) + this.targetStep.stepOffset;
                }
                else if (this.targetStep instanceof WizardStep) {
                    destinationStep = this.wizard.getIndexOfStep(this.targetStep);
                }
                else {
                    throw new Error("Input 'targetStep' is neither a WizardStep, StepOffset, StepIndex or StepId");
                }
                return destinationStep;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the `destinationStep`
         */
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the `destinationStep`
         * @param {?} event
         * @return {?}
         */
        GoToStepDirective.prototype.onClick = /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the `destinationStep`
         * @param {?} event
         * @return {?}
         */
        function (event) {
            this.wizard.goToStep(this.destinationStep, this.preFinalize, this.postFinalize);
        };
        GoToStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awGoToStep]'
                    },] }
        ];
        /** @nocollapse */
        GoToStepDirective.ctorParameters = function () { return [
            { type: WizardComponent },
            { type: WizardStep, decorators: [{ type: core.Optional }] }
        ]; };
        GoToStepDirective.propDecorators = {
            preFinalize: [{ type: core.Output }],
            postFinalize: [{ type: core.Output }],
            targetStep: [{ type: core.Input, args: ['awGoToStep',] }],
            finalize: [{ type: core.Output }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return GoToStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awNextStep` directive can be used to navigate to the next step.
     *
     * ### Syntax
     *
     * ```html
     * <button awNextStep (finalize)="finalize method">...</button>
     * ```
     *
     * @author Marc Arndt
     */
    var NextStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizard The state of the wizard
         */
        function NextStepDirective(wizard) {
            this.wizard = wizard;
            /**
             * This [[EventEmitter]] is called directly before the current step is exited during a transition through a component with this directive.
             */
            this.preFinalize = new core.EventEmitter();
            /**
             * This [[EventEmitter]] is called directly after the current step is exited during a transition through a component with this directive.
             */
            this.postFinalize = new core.EventEmitter();
        }
        Object.defineProperty(NextStepDirective.prototype, "finalize", {
            /**
             * A convenience field for `preFinalize`
             */
            get: /**
             * A convenience field for `preFinalize`
             * @return {?}
             */
            function () {
                return this.preFinalize;
            },
            /**
             * A convenience name for `preFinalize`
             *
             * @param emitter The [[EventEmitter]] to be set
             */
            set: /**
             * A convenience name for `preFinalize`
             *
             * @param {?} emitter The [[EventEmitter]] to be set
             * @return {?}
             */
            function (emitter) {
                /* istanbul ignore next */
                this.preFinalize = emitter;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the next step
         */
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the next step
         * @param {?} event
         * @return {?}
         */
        NextStepDirective.prototype.onClick = /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the next step
         * @param {?} event
         * @return {?}
         */
        function (event) {
            this.wizard.goToNextStep(this.preFinalize, this.postFinalize);
        };
        NextStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awNextStep]'
                    },] }
        ];
        /** @nocollapse */
        NextStepDirective.ctorParameters = function () { return [
            { type: WizardComponent }
        ]; };
        NextStepDirective.propDecorators = {
            preFinalize: [{ type: core.Output }],
            postFinalize: [{ type: core.Output }],
            finalize: [{ type: core.Output }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return NextStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awOptionalStep` directive can be used to define an optional `wizard-step`.
     * An optional wizard step is a [[WizardStep]] that doesn't need to be completed to transition to later wizard steps.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-step awOptionalStep>
     *     ...
     * </aw-wizard-step>
     * ```
     *
     * ### Example
     *
     * ```html
     * <aw-wizard-step stepTitle="Second step" awOptionalStep>
     *     ...
     * </aw-wizard-step>
     * ```
     *
     * @author Marc Arndt
     */
    var OptionalStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardStep The wizard step, which contains this [[OptionalStepDirective]]
         */
        function OptionalStepDirective(wizardStep) {
            this.wizardStep = wizardStep;
        }
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        OptionalStepDirective.prototype.ngOnInit = /**
         * Initialization work
         * @return {?}
         */
        function () {
            this.wizardStep.optional = true;
        };
        OptionalStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awOptionalStep]'
                    },] }
        ];
        /** @nocollapse */
        OptionalStepDirective.ctorParameters = function () { return [
            { type: WizardStep, decorators: [{ type: core.Host }] }
        ]; };
        return OptionalStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awPreviousStep` directive can be used to navigate to the previous step.
     * Compared to the [[NextStepDirective]] it's important to note, that this directive doesn't contain a `finalize` output method.
     *
     * ### Syntax
     *
     * ```html
     * <button awPreviousStep>...</button>
     * ```
     *
     * @author Marc Arndt
     */
    var PreviousStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizard The state of the wizard
         */
        function PreviousStepDirective(wizard) {
            this.wizard = wizard;
            /**
             * This [[EventEmitter]] is called directly before the current step is exited during a transition through a component with this directive.
             */
            this.preFinalize = new core.EventEmitter();
            /**
             * This [[EventEmitter]] is called directly after the current step is exited during a transition through a component with this directive.
             */
            this.postFinalize = new core.EventEmitter();
        }
        Object.defineProperty(PreviousStepDirective.prototype, "finalize", {
            /**
             * A convenience field for `preFinalize`
             */
            get: /**
             * A convenience field for `preFinalize`
             * @return {?}
             */
            function () {
                return this.preFinalize;
            },
            /**
             * A convenience field for `preFinalize`
             *
             * @param emitter The [[EventEmitter]] to be set
             */
            set: /**
             * A convenience field for `preFinalize`
             *
             * @param {?} emitter The [[EventEmitter]] to be set
             * @return {?}
             */
            function (emitter) {
                /* istanbul ignore next */
                this.preFinalize = emitter;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the previous step
         */
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the previous step
         * @param {?} event
         * @return {?}
         */
        PreviousStepDirective.prototype.onClick = /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the previous step
         * @param {?} event
         * @return {?}
         */
        function (event) {
            this.wizard.goToPreviousStep(this.preFinalize, this.postFinalize);
        };
        PreviousStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awPreviousStep]'
                    },] }
        ];
        /** @nocollapse */
        PreviousStepDirective.ctorParameters = function () { return [
            { type: WizardComponent }
        ]; };
        PreviousStepDirective.propDecorators = {
            preFinalize: [{ type: core.Output }],
            postFinalize: [{ type: core.Output }],
            finalize: [{ type: core.Output }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return PreviousStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awResetWizard` directive can be used to reset the wizard to its initial state.
     * This directive accepts an output, which can be used to specify some custom cleanup work during the reset process.
     *
     * ### Syntax
     *
     * ```html
     * <button awResetWizard (finalize)="custom reset task">...</button>
     * ```
     *
     * @author Marc Arndt
     */
    var ResetWizardDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizard The wizard component
         */
        function ResetWizardDirective(wizard) {
            this.wizard = wizard;
            /**
             * An [[EventEmitter]] containing some tasks to be done, directly before the wizard is being reset
             */
            this.finalize = new core.EventEmitter();
        }
        /**
         * Resets the wizard
         */
        /**
         * Resets the wizard
         * @param {?} event
         * @return {?}
         */
        ResetWizardDirective.prototype.onClick = /**
         * Resets the wizard
         * @param {?} event
         * @return {?}
         */
        function (event) {
            // do some optional cleanup work
            this.finalize.emit();
            // reset the wizard to its initial state
            this.wizard.reset();
        };
        ResetWizardDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awResetWizard]'
                    },] }
        ];
        /** @nocollapse */
        ResetWizardDirective.ctorParameters = function () { return [
            { type: WizardComponent }
        ]; };
        ResetWizardDirective.propDecorators = {
            finalize: [{ type: core.Output }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return ResetWizardDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awSelectedStep` directive can be used on a [[WizardStep]] to set it as selected after the wizard initialisation or a reset.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-step stepTitle="Step title" awSelectedStep>
     *     ...
     * </aw-wizard-step>
     * ```
     *
     * @author Marc Arndt
     */
    var SelectedStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardStep The wizard step, which should be selected by default
         */
        function SelectedStepDirective(wizardStep) {
            this.wizardStep = wizardStep;
        }
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        SelectedStepDirective.prototype.ngOnInit = /**
         * Initialization work
         * @return {?}
         */
        function () {
            this.wizardStep.defaultSelected = true;
        };
        SelectedStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awSelectedStep]'
                    },] }
        ];
        /** @nocollapse */
        SelectedStepDirective.ctorParameters = function () { return [
            { type: WizardStep, decorators: [{ type: core.Host }] }
        ]; };
        return SelectedStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awWizardCompletionStep` directive can be used to define a completion/success step at the end of your wizard
     * After a [[WizardCompletionStep]] has been entered, it has the characteristic that the user is blocked from
     * leaving it again to a previous step.
     * In addition entering a [[WizardCompletionStep]] automatically sets the `wizard`, and all steps inside the `wizard`,
     * as completed.
     *
     * ### Syntax
     *
     * ```html
     * <div awWizardCompletionStep [stepTitle]="title of the wizard step"
     *    [navigationSymbol]="{ symbol: 'navigation symbol', fontFamily: 'font-family' }"
     *    (stepEnter)="event emitter to be called when the wizard step is entered"
     *    (stepExit)="event emitter to be called when the wizard step is exited">
     *    ...
     * </div>
     * ```
     *
     * ### Example
     *
     * ```html
     * <div awWizardCompletionStep stepTitle="Step 1" [navigationSymbol]="{ symbol: '1' }">
     *    ...
     * </div>
     * ```
     *
     * With a navigation symbol from the `font-awesome` font:
     *
     * ```html
     * <div awWizardCompletionStep stepTitle="Step 1" [navigationSymbol]="{ symbol: '&#xf1ba;', fontFamily: 'FontAwesome' }">
     *    ...
     * </div>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardCompletionStepDirective = /** @class */ (function (_super) {
        __extends(WizardCompletionStepDirective, _super);
        function WizardCompletionStepDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WizardCompletionStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awWizardCompletionStep]',
                        providers: [
                            { provide: WizardStep, useExisting: core.forwardRef((/**
                                 * @return {?}
                                 */
                                function () { return WizardCompletionStepDirective; })) },
                            { provide: WizardCompletionStep, useExisting: core.forwardRef((/**
                                 * @return {?}
                                 */
                                function () { return WizardCompletionStepDirective; })) }
                        ]
                    },] }
        ];
        return WizardCompletionStepDirective;
    }(WizardCompletionStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awWizardStep` directive can be used to define a normal step inside a wizard.
     *
     * ### Syntax
     *
     * With `stepTitle` and `navigationSymbol` inputs:
     *
     * ```html
     * <div awWizardStep [stepTitle]="step title" [navigationSymbol]="{ symbol: 'symbol', fontFamily: 'font-family' }"
     *    [canExit]="deciding function" (stepEnter)="enter function" (stepExit)="exit function">
     *    ...
     * </div>
     * ```
     *
     * With `awWizardStepTitle` and `awWizardStepSymbol` directives:
     *
     * ```html
     * <div awWizardStep [canExit]="deciding function" (stepEnter)="enter function" (stepExit)="exit function">
     *    <ng-template awWizardStepTitle>
     *        step title
     *    </ng-template>
     *    <ng-template awWizardStepSymbol>
     *        symbol
     *    </ng-template>
     *    ...
     * </div>
     * ```
     *
     * ### Example
     *
     * With `stepTitle` and `navigationSymbol` inputs:
     *
     * ```html
     * <div awWizardStep stepTitle="Address information" [navigationSymbol]="{ symbol: '&#xf1ba;', fontFamily: 'FontAwesome' }">
     *    ...
     * </div>
     * ```
     *
     * With `awWizardStepTitle` and `awWizardStepSymbol` directives:
     *
     * ```html
     * <div awWizardStep>
     *    <ng-template awWizardStepTitle>
     *        Address information
     *    </ng-template>
     *    <ng-template awWizardStepSymbol>
     *        <i class="fa fa-taxi"></i>
     *    </ng-template>
     * </div>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardStepDirective = /** @class */ (function (_super) {
        __extends(WizardStepDirective, _super);
        function WizardStepDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WizardStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awWizardStep]',
                        providers: [
                            { provide: WizardStep, useExisting: core.forwardRef((/**
                                 * @return {?}
                                 */
                                function () { return WizardStepDirective; })) }
                        ]
                    },] }
        ];
        return WizardStepDirective;
    }(WizardStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The [[awNavigationMode]] directive can be used to customize wizard'd navigation mode.
     *
     * There are several usage options:
     *
     * ### Option 1. Customize the default navigation mode with [[navigateBackward]] and/or [[navigateForward]] inputs.
     *
     * ```html
     * <aw-wizard [awNavigationMode] navigateBackward="deny" navigateForward="allow">...</aw-wizard>
     * ```
     *
     * ### Option 2. Pass in a custom navigation mode
     *
     * ```typescript
     * import { BaseNavigationMode } from 'angular-archwizard'
     *
     * class CustomNavigationMode extends BaseNavigationMode {
     *
     *   // ...
     * }
     * ```
     *
     * ```typescript
     * \@Component({
     *   // ...
     * })
     * class MyComponent {
     *
     *   navigationMode = new CustomNavigationMode();
     * }
     * ```
     *
     * ```html
     * <aw-wizard [awNavigationMode]="navigationMode">...</aw-wizard>
     * ```
     *
     * ### Additional Notes
     *
     * - Specifying a custom navigation mode takes priority over [[navigateBackward]] and [[navigateForward]] inputs
     *
     * - Omitting the [[awNavigationMode]] directive or, equally, specifying just [[awNavigationMode]] without
     *   any inputs or parameters causes the wizard to use the default "strict" navigation mode equivalent to
     *
     * ```html
     * <aw-wizard [awNavigationMode] navigateBackward="deny" navigateForward="allow">...</aw-wizard>
     * ````
     */
    var NavigationModeDirective = /** @class */ (function () {
        function NavigationModeDirective(wizard) {
            this.wizard = wizard;
        }
        /**
         * @param {?} changes
         * @return {?}
         */
        NavigationModeDirective.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
        function (changes) {
            this.wizard.navigation = this.getNavigationMode();
        };
        /**
         * @private
         * @return {?}
         */
        NavigationModeDirective.prototype.getNavigationMode = /**
         * @private
         * @return {?}
         */
        function () {
            if (this.awNavigationMode) {
                return this.awNavigationMode;
            }
            return new ConfigurableNavigationMode(this.navigateBackward, this.navigateForward);
        };
        NavigationModeDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awNavigationMode]',
                    },] }
        ];
        /** @nocollapse */
        NavigationModeDirective.ctorParameters = function () { return [
            { type: WizardComponent }
        ]; };
        NavigationModeDirective.propDecorators = {
            awNavigationMode: [{ type: core.Input }],
            navigateBackward: [{ type: core.Input }],
            navigateForward: [{ type: core.Input }]
        };
        return NavigationModeDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awCompletedStep` directive can be used to make a wizard step initially completed.
     *
     * Initially completed steps are shown as completed when the wizard is presented to the user.
     *
     * A typical use case is to make a step initially completed if it is automatically filled with some derived/predefined information.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-step awCompletedStep>
     *     ...
     * </aw-wizard-step>
     * ```
     *
     * An optional boolean condition can be specified:
     *
     * ```html
     * <aw-wizard-step [awCompletedStep]="shouldBeCompleted">
     *     ...
     * </aw-wizard-step>
     * ```
     *
     * ### Example
     *
     * ```html
     * <aw-wizard-step stepTitle="First step" [awCompletedStep]="firstStepPrefilled">
     *     ...
     * </aw-wizard-step>
     * ```
     */
    var CompletedStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardStep The wizard step, which contains this [[CompletedStepDirective]]
         */
        function CompletedStepDirective(wizardStep) {
            this.wizardStep = wizardStep;
            // tslint:disable-next-line:no-input-rename
            this.initiallyCompleted = true;
        }
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        CompletedStepDirective.prototype.ngOnInit = /**
         * Initialization work
         * @return {?}
         */
        function () {
            // The input receives '' when specified in the template without a value.  In this case, apply the default value (`true`).
            this.wizardStep.initiallyCompleted = this.initiallyCompleted || (/** @type {?} */ (this.initiallyCompleted)) === '';
        };
        CompletedStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awCompletedStep]'
                    },] }
        ];
        /** @nocollapse */
        CompletedStepDirective.ctorParameters = function () { return [
            { type: WizardStep, decorators: [{ type: core.Host }] }
        ]; };
        CompletedStepDirective.propDecorators = {
            initiallyCompleted: [{ type: core.Input, args: ['awCompletedStep',] }]
        };
        return CompletedStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The module defining all the content inside `angular-archwizard`
     *
     * @author Marc Arndt
     */
    var ArchwizardModule = /** @class */ (function () {
        function ArchwizardModule() {
        }
        /* istanbul ignore next */
        /* istanbul ignore next */
        /**
         * @return {?}
         */
        ArchwizardModule.forRoot = /* istanbul ignore next */
        /**
         * @return {?}
         */
        function () {
            return {
                ngModule: ArchwizardModule,
                providers: [
                // Nothing here yet
                ]
            };
        };
        ArchwizardModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            WizardComponent,
                            WizardStepComponent,
                            WizardNavigationBarComponent,
                            WizardCompletionStepComponent,
                            GoToStepDirective,
                            NextStepDirective,
                            PreviousStepDirective,
                            OptionalStepDirective,
                            WizardStepSymbolDirective,
                            WizardStepTitleDirective,
                            EnableBackLinksDirective,
                            WizardStepDirective,
                            WizardCompletionStepDirective,
                            SelectedStepDirective,
                            ResetWizardDirective,
                            NavigationModeDirective,
                            CompletedStepDirective,
                        ],
                        imports: [
                            common.CommonModule
                        ],
                        exports: [
                            WizardComponent,
                            WizardStepComponent,
                            WizardNavigationBarComponent,
                            WizardCompletionStepComponent,
                            GoToStepDirective,
                            NextStepDirective,
                            PreviousStepDirective,
                            OptionalStepDirective,
                            WizardStepSymbolDirective,
                            WizardStepTitleDirective,
                            EnableBackLinksDirective,
                            WizardStepDirective,
                            WizardCompletionStepDirective,
                            SelectedStepDirective,
                            ResetWizardDirective,
                            NavigationModeDirective,
                            CompletedStepDirective,
                        ]
                    },] }
        ];
        return ArchwizardModule;
    }());

    exports.ArchwizardModule = ArchwizardModule;
    exports.BaseNavigationMode = BaseNavigationMode;
    exports.CompletedStepDirective = CompletedStepDirective;
    exports.ConfigurableNavigationMode = ConfigurableNavigationMode;
    exports.EnableBackLinksDirective = EnableBackLinksDirective;
    exports.GoToStepDirective = GoToStepDirective;
    exports.MovingDirection = MovingDirection;
    exports.NavigationModeDirective = NavigationModeDirective;
    exports.NextStepDirective = NextStepDirective;
    exports.OptionalStepDirective = OptionalStepDirective;
    exports.PreviousStepDirective = PreviousStepDirective;
    exports.ResetWizardDirective = ResetWizardDirective;
    exports.SelectedStepDirective = SelectedStepDirective;
    exports.WizardCompletionStep = WizardCompletionStep;
    exports.WizardCompletionStepComponent = WizardCompletionStepComponent;
    exports.WizardCompletionStepDirective = WizardCompletionStepDirective;
    exports.WizardComponent = WizardComponent;
    exports.WizardNavigationBarComponent = WizardNavigationBarComponent;
    exports.WizardStep = WizardStep;
    exports.WizardStepComponent = WizardStepComponent;
    exports.WizardStepDirective = WizardStepDirective;
    exports.WizardStepTitleDirective = WizardStepTitleDirective;
    exports.isStepId = isStepId;
    exports.isStepIndex = isStepIndex;
    exports.isStepOffset = isStepOffset;
    exports.a = WizardStepSymbolDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
